parlai/core/worlds.py

```diff
@@ -1499,6 +1533,622 @@ def shutdown(self):
+        batch.resized = resized
+        return batch
+
+    def enqueue_report(self):
+        self.report_queue.put_nowait(self.report())
+
+    def reset(self):
+        super().reset()
+        self.buffers = {}
+        self.num_prod = 0
+        self.num_consume = 0
+
+    def begin(self):
+        """
+        Signal world to begin.
+
+        World is ready to produce batches
+        """
+        self.begun = True
+
+
+class PBatchWorld(PWorld, BatchWorld):
+    """
+    P Batch World.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        world: Union[DialogPartnerWorld, MultiWorld],
+        produce_queue: Queue,
+        report_queue: Queue,
+        consume_queue: Queue,
+        worker_idx: int,
+    ):
+        PWorld.__init__(
+            self, opt, world, produce_queue, report_queue, consume_queue, worker_idx
+        )
+        BatchWorld.__init__(self, opt, world)
+
+    def produce_observation(self) -> Union[List[Message], Batch]:
+        teacher_idx = 0
+        agent_idx = 1
+        if hasattr(self.world, 'parley_init'):
+            for w in self.worlds:
+                w.parley_init()
+        batch_act = self.batch_act(teacher_idx, None)
+        batch_obs = self.batch_observe(agent_idx, batch_act, teacher_idx)
+        if hasattr(self.world.get_model_agent(), 'batchify'):
+            batch_obs = self._handle_buffers(batch_obs)
+        return batch_obs
+
+    def consume_act(self, act: List[Message]):
+        agent_idx = 1
+        for other_idx in range(len(self.world.get_agents())):
+            self.batch_observe(other_idx, act, agent_idx)
+
+    def reset(self):
+        PWorld.reset(self)
+        BatchWorld.reset(self)
+
+
+class PDynamicBatchWorld(PWorld, DynamicBatchWorld):
+    """
+    P Dynamic Batchworld.
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        world: Union[DialogPartnerWorld, MultiWorld],
+        produce_queue: Queue,
+        report_queue: Queue,
+        consume_queue: Queue,
+        worker_idx: int,
+    ):
+        PWorld.__init__(
+            self, opt, world, produce_queue, report_queue, consume_queue, worker_idx
+        )
+        DynamicBatchWorld.__init__(self, opt, world)
+        self.batch_indices: Optional[List[int]] = None
+
+    def produce_observation(self) -> Union[List[Message], Batch]:
+        assert self.batch_indices is None
+        batch, batch_obs = self._build_batch()
+        self.batch_indices = batch
+        if hasattr(self.world.get_model_agent(), 'batchify'):
+            batch_obs = self._handle_buffers(batch_obs)
+        return batch_obs
+
+    def consume_act(self, act: List[Message]):
+        assert self.batch_indices is not None
+        self._observe_batch_act(self.batch_indices, act)
+        self.batch_indices = None
+
+    def reset(self):
+        DynamicBatchWorld.reset(self)
+        PWorld.reset(self)
+        self.batch_indices = None
+
+
+def run_p_world(
```

Comment:
Without committing to a choice, some functions as large as this work well as Classes that implement __call__.

Also, should we consider making this private? I’m hoping we don’t ever import this anywhere else